File: notes/DSA/LeetCodeTop/438.py
import collections
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        need, window = collections.defaultdict(int), collections.defaultdict(int)
        for c in p:
            need[c] += 1
        
        left, right, vaild = 0, 0, 0
        s_len, p_len, need_len = len(s), len(p), len(need)
        res = []

        while right < s_len:
            c = s[right]
            right += 1

            if c in list(need.keys()):
                window[c] += 1
                if window[c] == need[c]:
                    vaild += 1

                while right - left >= p_len:
                    if vaild == need_len:
                        res.append(left)

                    d = s[left]
                    left += 1

                    if d in list(need.keys()):
                        if window[d] == need[d]:
                            vaild -= 1
                        window[d] -= 1
                    
        
        return res


if __name__ == "__main__":
    s = Solution()
    print(s.findAnagrams("acdcaeccde", "c"))

File: notes/DSA/LeetCodeTop/438.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string t) {
        unordered_map<char, int> need, window;
        for (char c : t) need[c]++;

        int left = 0, right = 0;
        int valid = 0;  // 记录条件
        int s_len = s.size(), t_len = t.size(), need_len = need.size();
        vector<int> res; // 记录结果

        while (right < s_len) {
            char c = s[right];
            right++;

            // 进行窗口内数据的一系列更新
            if (need.count(c)) {
                window[c]++;
                if (window[c] == need[c])
                    valid++;
            }

            // 判断左侧窗口是否要收缩
            while (right - left >= t_len) {
                // 当窗口符合条件时，把起始索引加入 res
                if (valid == need_len)
                    res.push_back(left);

                char d = s[left];
                left++;

                // 进行窗口内数据的一系列更新
                if (need.count(d)) {
                    if (window[d] == need[d])
                        valid--;
                    window[d]--;
                }
            }
        }

        return res;
    }
};


int main() {

    string s("acdcaeccde"), p("c");

    auto solution = new Solution;

    vector<int> res = solution->findAnagrams(s, p);
    for (auto r : res)
        std::cout << r << endl;
}